### PR TITLE
Create username on instantiation

### PR DESCRIPTION
**Pull request by glharper**
microsoft/cognitive-services-speech-sdk-js#405

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -135,6 +134,12 @@ const COMPRESSIONS = {
 class Cache {
   constructor(key, _) {
     const options = _ || {};
+    let tmpdir = path.join(os.tmpdir(), "async-disk-user");
+    try {
+      tmpdir = path.join(os.tmpdir(), username());
+    } catch (e) {
+      // use default tmpdir on username failure
```
Comment (stefanpenner) [inline index.js:141]: I have some questions:

0) if `username-sync` has an issue, maybe we should instead fix it there https://github.com/stefanpenner/username-sync
1) I would like to understand the failure mode here, could you explain?
2) in-order to prevent genuine bugs from sneaking in, my preference is for catch clauses to be explicitly targeted at specific errors (expected errors) , re-throwing the exceptional;

```js 

// my preferred pattern
} catch (e) {
  // let's ensure our catch clauses are specific
   if (e !== null && typeof e === 'object' && e.code === 'ENONT') {
      // comment describing why this is handled
   } else {
     throw e; // re-throw, because something unexpected happened
   }
}
```

File: index.js [hunk 2]
```diff
@@ -135,6 +134,12 @@ const COMPRESSIONS = {
 class Cache {
   constructor(key, _) {
     const options = _ || {};
+    let tmpdir = path.join(os.tmpdir(), "async-disk-user");
+    try {
+      tmpdir = path.join(os.tmpdir(), username());
```
Comment (stefanpenner) [inline index.js:139]: Running both `tmpdir()` and `username()` for each `Cache` created may slow down cache creation.

These values are not expected to change once a process starts, which is why they where originally derived in module scope above. Was that change needed?

**Review by stefanpenner — CHANGES_REQUESTED** [review]

File: index.js (hunk 1 above)
Comment (glharper) [reply index.js:141]: I do have an open PR for username-sync [here](https://github.com/stefanpenner/username-sync/pull/4)

Comment (stefanpenner) [reply index.js:141]: @glharper oo, i didn't notice that. I much prefer that PR. Mind if we close this one, and work to land that one?

**Comment by stefanpenner** [comment]
Closing in-favor of https://github.com/stefanpenner/username-sync/pull/4 which mitigates the problem in a common place and as a bonus also addresses [sync-disk-cache](https://github.com/stefanpenner/sync-disk-cache) and any other users.

Thank you @glharper!
